File: app/src/main/java/com/soundcloud/android/utils/collection/AsyncLoader.java
package com.soundcloud.android.utils.collection;

import static com.soundcloud.android.transformers.Transformers.doOnFirst;

import com.soundcloud.android.model.AsyncLoadingState;
import com.soundcloud.android.view.ViewError;
import com.soundcloud.java.optional.Optional;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.BiFunction;
import io.reactivex.functions.Function;
import io.reactivex.observables.ConnectableObservable;
import io.reactivex.subjects.BehaviorSubject;

import android.support.annotation.NonNull;

import javax.inject.Provider;
import java.util.ArrayList;
import java.util.List;

public class AsyncLoader<PageData, FirstPageParamsType> extends Observable<AsyncLoaderState<PageData>> {

    private final Function<FirstPageParamsType, Observable<PageResult<PageData>>> paramsToFirstPage;
    private final Function<FirstPageParamsType, Observable<PageResult<PageData>>> paramsToRefresh;
    private final Observable<FirstPageParamsType> refreshRequested;
    private final Observable<FirstPageParamsType> firstPageRequested;
    private final Observable<Object> nextPageRequested;
    private final Optional<BiFunction<PageData, PageData, PageData>> pageCombiner;
    private final CompositeDisposable compositeDisposable = new CompositeDisposable();

    private final BehaviorSubject<Optional<Provider<Observable<PageResult<PageData>>>>> nextPage = BehaviorSubject.create();
    private final BehaviorSubject<RefreshState> refreshStateSubject = BehaviorSubject.createDefault(new RefreshState(false, Optional.absent()));

    public static class PageResult<PageData> {

        PageData data;
        Optional<Provider<Observable<PageResult<PageData>>>> nextPage;

        PageResult(PageData data, Optional<Provider<Observable<PageResult<PageData>>>> nextPage) {
            this.data = data;
            this.nextPage = nextPage;
        }

        public static <PageData> PageResult<PageData> from(PageData data, Provider<Observable<PageResult<PageData>>> nextPage) {
            return from(data, Optional.of(nextPage));
        }

        public static <PageData> PageResult<PageData> from(PageData data) {
            return from(data, Optional.absent());
        }

        public static <PageData> PageResult<PageData> from(PageData data, Optional<Provider<Observable<PageResult<PageData>>>> nextPage) {
            return new PageResult<>(data, nextPage);
        }
    }

    private final class RefreshState {

        private final boolean loading;
        private final Optional<Throwable> error;

        RefreshState(boolean loading, Optional<Throwable> error) {
            this.loading = loading;
            this.error = error;
        }

    }

    private final class PageState {

        private final Optional<PageData> data;
        private final boolean loading;
        private final Optional<Throwable> error;

        PageState(Optional<PageData> data, boolean loading, Optional<Throwable> error) {
            this.data = data;
            this.loading = loading;
            this.error = error;
        }
    }

    AsyncLoader(@NonNull Observable<FirstPageParamsType> firstPageRequested,
                @NonNull Function<FirstPageParamsType, Observable<PageResult<PageData>>> paramsToFirstPage,
                @NonNull Observable<FirstPageParamsType> refreshRequested,
                @NonNull Function<FirstPageParamsType, Observable<PageResult<PageData>>> paramsToRefresh,
                @NonNull Observable<Object> nextPageRequested,
                Optional<BiFunction<PageData, PageData, PageData>> pageCombiner) {
        this.paramsToRefresh = paramsToRefresh;
        this.paramsToFirstPage = paramsToFirstPage;
        this.refreshRequested = refreshRequested;
        this.firstPageRequested = firstPageRequested;
        this.nextPageRequested = nextPageRequested;
        this.pageCombiner = pageCombiner;
    }


    @Override
    protected void subscribeActual(Observer<? super AsyncLoaderState<PageData>> observer) {
        Observable<Observable<PageState>> sequenceStarters = refreshes().startWith(initialLoad());
        Observable.combineLatest(sequenceStarters.switchMap(this::createSequenceState),
                                        refreshStateSubject,
                                        this::updateWithRefresh)
                         .doOnDispose(this::cleanup)
                .subscribe(observer);
    }

    private void cleanup() {
        compositeDisposable.dispose();
    }


    private Observable<Observable<PageState>> refreshes() {
        return refreshRequested
                .map(params -> {
                    ConnectableObservable<PageResult<PageData>> replay = paramsToRefresh.apply(params).replay(1);
                    compositeDisposable.add(replay.connect());
                    return replay;
                })
                .flatMap(refreshObservable -> Observable.create(e -> refreshObservable.take(1)
                                                                                      .doOnSubscribe(disposable -> refreshStateSubject.onNext(new RefreshState(true, Optional.absent())))
                                                                                      .doOnError(throwable -> refreshStateSubject.onNext(new RefreshState(false, Optional.of(throwable))))
                                                                                      .doOnNext(p -> refreshStateSubject.onNext(new RefreshState(false, Optional.absent())))
                                                                                      .subscribe(pageState -> e.onNext(refreshObservable.lift(doOnFirst(this::keepNextPageObservable))
                                                                                                                                        .map(AsyncLoader.this::loadedPageState)))));
    }

    private Observable<AsyncLoaderState<PageData>> createSequenceState(Observable<PageState> firstPage) {
        return pageEmitter().startWith(firstPage)
                            .scan(new ArrayList<>(), this::addNewPage)
                            .switchMap(this::stateFromPages);
    }

    private AsyncLoaderState<PageData> updateWithRefresh(AsyncLoaderState<PageData> pageDataAsyncLoaderState, RefreshState refreshState) {
        return pageDataAsyncLoaderState.updateWithRefreshState(refreshState.loading, refreshState.error);
    }

    @NonNull
    private List<Observable<PageState>> addNewPage(List<Observable<PageState>> observables, Observable<PageState> partialStateObservable) {
        observables.add(replayLastAndConnect(partialStateObservable));
        return observables;
    }

    @NonNull
    private ConnectableObservable<PageState> replayLastAndConnect(Observable<PageState> partialStateObservable) {
        ConnectableObservable<PageState> replay = partialStateObservable.replay(1);
        compositeDisposable.add(replay.connect());
        return replay;
    }

    private Observable<AsyncLoaderState<PageData>> stateFromPages(List<Observable<PageState>> arr) {
        return Observable.combineLatest(arr, this::combinePages);
    }

    private AsyncLoaderState<PageData> combinePages(Object[] objects) throws Exception {
        PageData combinedData = combinePageData(objects);
        PageState lastPageState = lastPageState(objects);
        AsyncLoadingState loadingState = AsyncLoadingState.builder()
                                                          .isLoadingNextPage(lastPageState.loading)
                                                          .requestMoreOnScroll(!lastPageState.loading && !lastPageState.error.isPresent() && hasMorePages())
                                                          .nextPageError(lastPageState.error.transform(ViewError::from))
                                                          .build();

        return AsyncLoaderState.<PageData>builder().data(Optional.fromNullable(combinedData))
                                                   .asyncLoadingState(loadingState)
                                                   .build();
    }

    private PageData combinePageData(Object[] objects) throws Exception {
        PageData combinedData = null;
        for (Object object : objects) {
            PageState partialState = (PageState) object;
            if (partialState.data.isPresent()) {
                if (combinedData == null) {
                    combinedData = partialState.data.get();
                } else {
                    combinedData = pageCombiner.get().apply(combinedData, partialState.data.get());
                }
            }
        }
        return combinedData;
    }

    private PageState lastPageState(Object[] objects) {
        return (PageState) objects[objects.length - 1];
    }

    @NonNull
    private Observable<Observable<PageState>> pageEmitter() {
        return nextPageRequested.withLatestFrom(nextPage, (__, nextPage1) -> nextPage1)
                                .filter(Optional::isPresent)
                                .map(nextPageOpt -> nextPageOpt.get().get())
                                .map(this::nextPageObservable);
    }

    private Observable<PageState> initialLoad() {
        return firstPageRequested.flatMap(params -> paramsToFirstPage.apply(params)
                                                                     .lift(doOnFirst(this::keepNextPageObservable))
                                                                     .map(this::loadedPageState)
                                                                     .onErrorReturn(this::errorPageState)
                                                                     .startWith(loadingPageState()));
    }

    @NonNull
    private PageState errorPageState(Throwable t) {
        return new PageState(Optional.absent(), false, Optional.of(t));
    }

    private Observable<PageState> nextPageObservable(Observable<PageResult<PageData>> nextPage) {
        return nextPage.lift(doOnFirst(this::keepNextPageObservable))
                       .map(this::loadedPageState)
                       .onErrorReturn(this::errorPageState)
                       .startWith(loadingPageState());
    }

    @NonNull
    private PageState loadingPageState() {
        return new PageState(Optional.absent(), true, Optional.absent());
    }

    @NonNull
    private PageState loadedPageState(PageResult<PageData> result) {
        return new PageState(Optional.of(result.data), false, Optional.absent());
    }

    private void keepNextPageObservable(PageResult<PageData> datas) {
        try {
            nextPage.onNext(datas.nextPage);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private boolean hasMorePages() {
        return nextPage.hasValue() && nextPage.getValue().isPresent();
    }

    public static <PageData, FirstPageParamsType> Builder<PageData, FirstPageParamsType> startWith(Observable<FirstPageParamsType> initialLoadSignal,
                                                                                                   Function<FirstPageParamsType, Observable<PageResult<PageData>>> loadInitialPageWith) {
        return new Builder<>(initialLoadSignal, loadInitialPageWith);
    }

    public static class Builder<PageData, FirstPageParamsType> {
        private final Observable<FirstPageParamsType> firstPageRequested;
        private final Function<FirstPageParamsType, Observable<PageResult<PageData>>> firstPage;
        private Observable<FirstPageParamsType> refreshRequested = Observable.never();
        private Function<FirstPageParamsType, Observable<PageResult<PageData>>> refreshWith = __ -> Observable.empty();
        private Observable<Object> nextPageRequested = Observable.never();
        private Optional<BiFunction<PageData, PageData, PageData>> combinerOpt = Optional.absent();

        public Builder(Observable<FirstPageParamsType> initialLoadSignal,
                       Function<FirstPageParamsType, Observable<PageResult<PageData>>> paramsToLoadOp) {
            this.firstPageRequested = initialLoadSignal;
            this.firstPage = paramsToLoadOp;
        }

        public Builder<PageData, FirstPageParamsType> withRefresh(Observable<FirstPageParamsType> refreshSignal,
                                                                  Function<FirstPageParamsType, Observable<PageResult<PageData>>> paramsToRefreshOp) {
            this.refreshRequested = refreshSignal;
            this.refreshWith = paramsToRefreshOp;
            return this;
        }

        public Builder<PageData, FirstPageParamsType> withPaging(Observable<Object> nextPageSignal,
                                                                 BiFunction<PageData, PageData, PageData> pageCombiner) {
            this.nextPageRequested = nextPageSignal;
            this.combinerOpt = Optional.of(pageCombiner);
            return this;
        }

        public AsyncLoader<PageData, FirstPageParamsType> build() {
            return new AsyncLoader<>(
                    firstPageRequested,
                    firstPage,
                    refreshRequested,
                    refreshWith,
                    nextPageRequested,
                    combinerOpt
            );
        }
    }
}

File: app/src/main/java/com/soundcloud/android/utils/collection/AsyncLoader.kt
package com.soundcloud.android.utils.collection

import com.soundcloud.android.model.AsyncLoadingState
import com.soundcloud.android.transformers.Transformers.doOnFirst
import com.soundcloud.android.view.ViewError
import com.soundcloud.java.optional.Optional
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observables.ConnectableObservable
import io.reactivex.rxkotlin.Observables
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.subjects.BehaviorSubject
import javax.inject.Provider

class AsyncLoader<PageData, FirstPageParamsType> internal constructor(private val firstPageRequested: Observable<FirstPageParamsType>,
                                                                                  private val paramsToFirstPage: (FirstPageParamsType) -> Observable<PageResult<PageData>>,
                                                                                  private val refreshRequested: Observable<FirstPageParamsType>,
                                                                                  private val paramsToRefresh: (FirstPageParamsType) -> Observable<PageResult<PageData>>,
                                                                                  private val nextPageRequested: Observable<Any>,
                                                                                  private val pageCombiner: Optional<(PageData, PageData) -> PageData>)
    : Observable<AsyncLoaderState<PageData>>() {

    private val compositeDisposable = CompositeDisposable()
    private val nextPage = BehaviorSubject.create<Optional<Provider<Observable<PageResult<PageData>>>>>()
    private val refreshStateSubject = BehaviorSubject.createDefault(RefreshState(false, Optional.absent<Throwable>()))

    class PageResult<PageData> internal constructor(internal var data: PageData,
                                                                internal var nextPage: Optional<Provider<Observable<PageResult<PageData>>>> = Optional.absent()) {
        companion object {

            fun <PageData> from(data: PageData, nextPage: Provider<Observable<PageResult<PageData>>>): PageResult<PageData> = PageResult(data, nextPage = Optional.of(nextPage))

            fun <PageData> from(data: PageData): PageResult<PageData> = PageResult(data)

            fun <PageData> from(data: PageData, nextPage: Optional<Provider<Observable<PageResult<PageData>>>>): PageResult<PageData> = PageResult(data, nextPage = nextPage)
        }
    }

    private data class RefreshState internal constructor(internal val loading: Boolean, internal val error: Optional<Throwable> = Optional.absent())

    private inner class PageState internal constructor(internal val data: Optional<PageData> = Optional.absent(),
                                                       internal val loading: Boolean = false,
                                                       internal val error: Optional<Throwable> = Optional.absent())

    override fun subscribeActual(observer: Observer<in AsyncLoaderState<PageData>>) {
        val sequenceStarters = refreshes().startWith(initialLoad())

        Observables.combineLatest(sequenceStarters.switchMap { createSequenceState(it) },
                                       refreshStateSubject,
                                       { pageDataAsyncLoaderState, refreshState -> updateWithRefresh(pageDataAsyncLoaderState, refreshState)})
                .doOnDispose { compositeDisposable.clear() }
                .subscribe(observer)
    }

    private fun refreshes(): Observable<Observable<PageState>> {
        return refreshRequested
                .map { params ->
                    replayLastAndConnect(paramsToRefresh(params))
                }
                .flatMap { refreshObservable ->
                    Observable.create<Observable<PageState>> { e ->
                        compositeDisposable.add(refreshObservable.take(1)
                                                        .doOnSubscribe { refreshStateSubject.onNext(RefreshState(true)) }
                                                        .subscribeBy(onNext = {
                                                            refreshStateSubject.onNext(RefreshState(false))
                                                            e.onNext(refreshObservable.lift(doOnFirst(this::keepNextPageObservable)).map { loadedPageState(it) })
                                                        },
                                                                     onError = { throwable -> refreshStateSubject.onNext(RefreshState(false, Optional.of(throwable))) }))
                    }
                }
    }

    private fun createSequenceState(firstPage: Observable<PageState>): Observable<AsyncLoaderState<PageData>> {
        return pageEmitter().startWith(firstPage)
                .scan(mutableListOf<Observable<PageState>>(), { observables, partialStateObservable -> this.addNewPage(observables, partialStateObservable) })
                .switchMap { this.stateFromPages(it) }
    }

    private fun updateWithRefresh(pageDataAsyncLoaderState: AsyncLoaderState<PageData>, refreshState: RefreshState): AsyncLoaderState<PageData> =
            pageDataAsyncLoaderState.updateWithRefreshState(refreshState.loading, refreshState.error)

    private fun addNewPage(observables: MutableList<Observable<PageState>>, partialStateObservable: Observable<PageState>): MutableList<Observable<PageState>> {
        observables.add(replayLastAndConnect(partialStateObservable))
        return observables
    }

    private fun <T> replayLastAndConnect(partialStateObservable: Observable<T>): ConnectableObservable<T> {
        val replay = partialStateObservable.replay(1)
        compositeDisposable.add(replay.connect())
        return replay
    }

    private fun stateFromPages(arr: List<Observable<PageState>>): Observable<AsyncLoaderState<PageData>> = Observable.combineLatest(arr) { this.combinePages(it) }

    @Throws(Exception::class)
    private fun combinePages(objects: Array<Any>): AsyncLoaderState<PageData> {
        val combinedData = combinePageData(objects)
        val lastPageState = toPageState(objects.last())
        val loadingState = AsyncLoadingState.builder()
                .isLoadingNextPage(lastPageState.loading)
                .requestMoreOnScroll(!lastPageState.loading && !lastPageState.error.isPresent && hasMorePages())
                .nextPageError(lastPageState.error.transform { ViewError.from(it) })
                .build()

        return if (combinedData != null) AsyncLoaderState(data = Optional.of(combinedData),
                                                          asyncLoadingState = loadingState)
        else AsyncLoaderState(asyncLoadingState = loadingState)
    }

    @Throws(Exception::class)
    private fun combinePageData(objects: Array<Any>): PageData? {
        val empty: PageData? = null
        return objects.map { toPageState(it) }
                .filter { it.data.isPresent }.fold(empty) { acc, pageState ->
            if (acc == null) {
                pageState.data.get()
            } else {
                pageCombiner.get()(acc, pageState.data.get())
            }
        }
    }

    @Suppress("UNCHECKED_CAST", "UnsafeCast")
    private fun toPageState(any: Any): PageState = any as AsyncLoader<PageData, FirstPageParamsType>.PageState

    private fun pageEmitter(): Observable<Observable<PageState>> {
        return nextPageRequested.flatMap { nextPage.take(1) }
                .filter { it.isPresent }
                .map { nextPageOpt -> nextPageOpt.get().get() }
                .map { this.nextPageObservable(it) }
    }

    private fun initialLoad(): Observable<PageState> {
        return firstPageRequested.flatMap { params ->
            paramsToFirstPage(params)
                    .lift(doOnFirst<PageResult<PageData>>(this::keepNextPageObservable))
                    .map { this.loadedPageState(it) }
                    .onErrorReturn { this.errorPageState(it) }
                    .startWith(loadingPageState())
        }
    }

    private fun errorPageState(t: Throwable): PageState = PageState(error = Optional.of(t))

    private fun nextPageObservable(nextPage: Observable<PageResult<PageData>>): Observable<PageState> {
        return nextPage.lift(doOnFirst<PageResult<PageData>>(this::keepNextPageObservable))
                .map { this.loadedPageState(it) }
                .onErrorReturn { this.errorPageState(it) }
                .startWith(loadingPageState())
    }

    private fun loadingPageState(): PageState = PageState(loading = true)

    private fun loadedPageState(result: PageResult<PageData>): PageState = PageState(data = Optional.of(result.data))

    private fun keepNextPageObservable(datas: PageResult<PageData>) {
        nextPage.onNext(datas.nextPage)
    }

    private fun hasMorePages(): Boolean = nextPage.hasValue() && nextPage.value.isPresent

    class Builder<PageData, FirstPageParamsType>(private val firstPageRequested: Observable<FirstPageParamsType>,
                                                             private val firstPage: (FirstPageParamsType) -> Observable<PageResult<PageData>>) {
        private var refreshRequested = Observable.never<FirstPageParamsType>()
        private var refreshWith: (FirstPageParamsType) -> Observable<PageResult<PageData>> = { empty() }
        private var nextPageRequested = Observable.never<Any>()
        private var combinerOpt = Optional.absent<(PageData, PageData) -> PageData>()

        fun withRefresh(refreshSignal: Observable<FirstPageParamsType>,
                        paramsToRefreshOp: (FirstPageParamsType) -> Observable<PageResult<PageData>>): Builder<PageData, FirstPageParamsType> {
            this.refreshRequested = refreshSignal
            this.refreshWith = paramsToRefreshOp
            return this
        }

        fun withPaging(nextPageSignal: Observable<Any>,
                       pageCombiner: (PageData, PageData) -> PageData): Builder<PageData, FirstPageParamsType> {
            this.nextPageRequested = nextPageSignal
            this.combinerOpt = Optional.of(pageCombiner)
            return this
        }

        fun build(): AsyncLoader<PageData, FirstPageParamsType> {
            return AsyncLoader(
                    firstPageRequested,
                    firstPage,
                    refreshRequested,
                    refreshWith,
                    nextPageRequested,
                    combinerOpt
            )
        }
    }

    companion object {

        @JvmStatic
        fun <PageData, FirstPageParamsType> startWith(initialLoadSignal: Observable<FirstPageParamsType>,
                                                                  loadInitialPageWith: (FirstPageParamsType) -> Observable<PageResult<PageData>>):
                Builder<PageData, FirstPageParamsType> = Builder(initialLoadSignal, loadInitialPageWith)
    }
}
